scanner/src/main/kotlin/experimental/ProvenanceBasedFileStorage.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2021 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.experimental
+
+import com.fasterxml.jackson.module.kotlin.readValue
+
+import java.io.ByteArrayInputStream
+import java.io.FileNotFoundException
+import java.io.IOException
+
+import org.ossreviewtoolkit.model.ArtifactProvenance
+import org.ossreviewtoolkit.model.KnownProvenance
+import org.ossreviewtoolkit.model.RepositoryProvenance
+import org.ossreviewtoolkit.model.ScanResult
+import org.ossreviewtoolkit.model.yamlMapper
+import org.ossreviewtoolkit.scanner.ScannerCriteria
+import org.ossreviewtoolkit.utils.collectMessagesAsString
+import org.ossreviewtoolkit.utils.fileSystemEncode
+import org.ossreviewtoolkit.utils.log
+import org.ossreviewtoolkit.utils.showStackTrace
+import org.ossreviewtoolkit.utils.storage.FileStorage
+
+class ProvenanceBasedFileStorage(private val backend: FileStorage) : ProvenanceBasedScanStorage {
+    override fun read(provenance: KnownProvenance): List<ScanResult> {
+        requireVcsPathBlank(provenance)
+
+        val path = storagePath(provenance)
+
+        @Suppress("TooGenericExceptionCaught")
+        return try {
+            backend.read(path).use { input ->
+                yamlMapper.readValue<List<ScanResult>>(input).map {
+                    // Use the provided provenance for the result instead of building it from the stored values, because
+                    // in the case of a RepositoryRevision only the resolved revision matters.
+                    it.copy(provenance = provenance)
+                }
+            }
+        } catch (e: Exception) {
+            when (e) {
+                is FileNotFoundException -> {
+                    // If the file cannot be found it means no scan results have been stored, yet.
+                    emptyList()
+                }
+                else -> {
+                    val message = "Could not read scan results for '$provenance' from path '$path': " +
+                            e.collectMessagesAsString()
+
+                    log.info { message }
+
+                    // TODO: Propagate error.
+                    emptyList()
+                }
+            }
+        }
+    }
+
+    override fun read(provenance: KnownProvenance, scannerCriteria: ScannerCriteria): List<ScanResult> =
```

Comment:
I think, this function can now be removed, as it is defined in the `ProvenanceBasedScanStorage` interface. The same is true for the Postgres-based implementation.